src/video/video_driver.cpp

```diff
@@ -15,69 +15,129 @@
+			 * fast-forwarding. We have to check if the draw-thread wants to
+			 * gain access to the game-state, and give him a chance to get the
+			 * lock, as otherwise we would be keeping the lock constantly.
+			 * Without the sleep_for(), we would normally not be yielding.
+			 * So, we check if the draw-thread wants to have the game-state
+			 * lock, and if so give it a chance to get it, by waiting for the
+			 * draw-thread to inform us it got the lock.
+			 */
+			if (this->request_game_state_mutex) {
+				std::mutex wait_mutex;
+				std::lock_guard<std::mutex> wait_lock(wait_mutex);
+				this->game_state_locked_signal.wait(wait_mutex);
+			}
 		}
+	}
+}
 
-		/* The game loop is the part that can run asynchronously.
-		 * The rest except sleeping can't. */
-		this->UnlockVideoBuffer();
-		::GameLoop();
-		this->LockVideoBuffer();
+/* static */ void VideoDriver::GameThreadThunk(VideoDriver *drv)
+{
+	drv->GameThread();
+}
+
+void VideoDriver::StartGameThread()
+{
+	if (this->is_game_threaded) {
+		this->is_game_threaded = StartNewThread(nullptr, "ottd:game", &VideoDriver::GameThreadThunk, this);
+	}
+
+	DEBUG(driver, 1, "using %sthread for game-loop", this->is_game_threaded ? "" : "no ");
+}
+
+void VideoDriver::Tick()
+{
+	auto now = std::chrono::steady_clock::now();
+
+	/* If more than a millisecond has passed, increase the _realtime_tick. */
+	if (now - this->last_realtime_tick >= std::chrono::milliseconds(1)) {
+		auto delta = std::chrono::duration_cast<std::chrono::milliseconds>(now - this->last_realtime_tick);
+		_realtime_tick += delta.count();
+		this->last_realtime_tick += delta;
+	}
+
+	if (!this->is_game_threaded && now >= this->next_game_tick) {
+		this->GameLoop();
 
 		/* For things like dedicated server, don't run a separate draw-tick. */
 		if (!this->HasGUI()) {
 			::InputLoop();
-			UpdateWindows();
+			::UpdateWindows();
 			this->next_draw_tick = this->next_game_tick;
 		}
 	}
 
-	/* Prevent drawing when switching mode, as windows can be removed when they should still appear. */
-	if (this->HasGUI() && cur_ticks >= this->next_draw_tick && (_switch_mode == SM_NONE || HasModalProgress())) {
+	if (this->HasGUI() && now >= this->next_draw_tick) {
 		this->next_draw_tick += this->GetDrawInterval();
 		/* Avoid next_draw_tick getting behind more and more if it cannot keep up. */
-		if (this->next_draw_tick < cur_ticks - ALLOWED_DRIFT * this->GetDrawInterval()) this->next_draw_tick = cur_ticks;
+		if (this->next_draw_tick < now - ALLOWED_DRIFT * this->GetDrawInterval()) this->next_draw_tick = now;
 
 		this->InputLoop();
-		::InputLoop();
-		UpdateWindows();
+
+		{
+			/* Inform the game-thread that we want the game-state, get the
+			 * lock on the game-state, and inform the game-thread that we
+			 * got the lock. This avoids the game-thread to never yield.
+			 * See comment in the game-thread for more details. */
+			this->request_game_state_mutex = true;
+			std::lock_guard<std::mutex> lock(this->game_state_mutex);
```

Comment:
I don't think this works. Example:
1. Tick(): request_game_state_mutex = true
2. Gameloop(): game_state_mutex.unlock()
3. Tick(): game_state_mutex.lock()
4. Gameloop(): request_game_state_mutex.load()
5. Tick(): notify_one(), but noone is waiting
6. Tick(): request_game_state_mutex = false
7. Gameloop(): waiting for condition_variable, but noone notifies

Instead you may try to use two mutexes: game_state_mutex, wait_mutex
* GameThead(): locks/unlocks wait_mutex between GameLoops. So this thread alternates between locking game_state_mutex and wait_mutex, never both.
* Tick(): Lock wait_mutex, then lock game_state_mutex.